File: JavaDefendYourCode/src/utils/FileUtil.java
package utils;

import logging.MyLogger;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;
import java.util.logging.Level;

/*
Team: Abusement Park 3.0
Members: Kenny White, Jordan Lambert, Daric Sage
 */

public class FileUtil {

    public static File getFile(Scanner kb, boolean isInput){

        String fileType = (isInput) ? "input" : "output";
        String display = "Please enter name of the " + fileType + " file(must be in local directory): ";

        String fileName = InputUtil.stringInput(kb, display);

        boolean valid = (isInput) ? FileUtil.isValidInputFile(fileName) : FileUtil.isValidOutputFile(fileName);

        while(!valid){

            System.out.println("That is not a valid file name, or not in the local directory, or cannot be opened, please retry.");

            fileName = InputUtil.stringInput(kb, display);

            valid = (isInput) ? FileUtil.isValidInputFile(fileName) : FileUtil.isValidOutputFile(fileName);
        }
        return new File(fileName);
    }


    private static boolean isValidInputFile(String fileName){
        try {
            File fh = new File(fileName);
            return  fh.exists() &&
                    !fh.isDirectory() &&
                    !fh.isAbsolute() &&
                    fh.canRead();
        }
        catch (Exception ex){
            return false;
        }
    }
    private static boolean isValidOutputFile(String fileName){
        try {
            File fh = new File(fileName);
            boolean valid = true;
            if(fh.exists()){
                valid = fh.canWrite();
            }
            return valid && !fh.isDirectory() && !fh.isAbsolute();

        }
        catch (Exception ex){
            return false;
        }
    }


    public static void writeInputToOutput(File input, PrintWriter output){
        try{
            Scanner fin = new Scanner(input);
            while (fin.hasNextLine()){
                output.println(fin.nextLine());
            }
            fin.close();
        }
        catch (Exception ex){
            System.out.println("There was a problem writing from input to output file.");
            MyLogger.log(Level.SEVERE, "Failed to copy input file to output: \n\t{0}", ex.getMessage());
        }
    }

    public static PrintWriter getPrintWriter(File outputFH) throws IOException{
        try{
            return new PrintWriter(outputFH);
        }
        catch (Exception ex){
            MyLogger.log(Level.SEVERE, "Cannot open output PrintWriter\n\t{0}\n", outputFH, ex.getMessage());
            throw new IOException("Failed to open PrintWriter");
        }
    }

}

File: CDefendYourCode/utils/fileUtil.c

#include "fileUtil.h"

/*
Team: Abusement Park 3.0
Members: Kenny White, Jordan Lambert, Daric Sage
 */

void getOutputPath(char* path){

    int result = 0;

    while(!result)
    {
        printf("\n%s", "Please enter name of output file (file must be in current directory and you can only enter the file name."
                " \nNot a direct path. Characters can be [a-zA-Z0-9\\-.] 1 to 20 chars and must end in .txt)\n"
                "WARNING: Output file will be overwritten if it exists.\n");
        fgets(path, LINE_SIZE, stdin);
        result = checkFile(path, 1);
    }

}

void getInputPath(char* path) {

    int result= 0;

    while(!result){
        printf("\n%s", "Please enter name of input file (file must be in current directory and you can only enter the file name."
                " \nNot a direct path. Characters can be [a-zA-Z0-9-.]) 1 to 20 chars and must end in .txt)\n");
        fgets(path, LINE_SIZE, stdin);
        result = checkFile(path, 0);
    }

}



int checkFile(char * filename, int isOutput)
{
    size_t i;
    char flags[20] = "r";
    if(isOutput) {
        bzero(flags, 20);
        strncpy(flags, "w+\0", 3);
    }
    size_t length = strnlen(filename, LINE_SIZE);
    for(i = 0 ; i < length ; i++)
    {
        if(filename[i] == '\n' || i == (LINE_SIZE-1)) {
            filename[i] = '\0';
            break;
        }
    }

    compileFileRegex();
    int isvalid = regexIsValid(filename);
    if(isvalid)
    {
        char resolvedpath[LINE_SIZE];
        realpath(filename, resolvedpath);
        FILE * fp;
        fp = fopen(resolvedpath,flags);
        if (fp != NULL)
        {
            fclose(fp);
            return 1;
        }
        else
        {
            printf("Not a valid file, please retry.\n");
            return 0;
        }
    }

    printf("File validation failed, please retry.\n");
    return 0;
}


FILE* openFileRead(const char *path) {
    FILE * fp;
    fp = fopen (path, "r");
    return fp;
}

FILE* openFileWrite(const char *path) {
    FILE * fp;
    fp = fopen (path, "w+");
    return fp;
}

void writeToErrorFile(char message[]){
    FILE* fp = fopen("ErrorLog.txt", "aw+");
    if(fp != NULL){
        fprintf(fp, "\n%s", message );
        fclose(fp);
    }
    else{
        fprintf(stderr, "\nFailed to open ErrorLog.txt to record error:\n%s\n", message);
    }
}

void writeInputToOutput(FILE* fin, FILE* fout){
    char line[100];
    while(fgets(line, sizeof(line), fin) !=NULL){
        fputs(line, fout);
    }
    printf("\nWrote input to output file.\n");
}

off_t fileSize(int fd){

    struct stat st;
    fstat(fd, &st);
    return st.st_size;
}